service/src/main/java/uk/nhs/adaptors/oneoneone/cda/report/mapper/ConsentMapper.java

```diff
@@ -0,0 +1,190 @@
+            narrative.setDivAsString(section.getText().xmlText());
+            consent.setText(narrative);
+        }
+    }
+
+    private void extractConsentSource(Consent consent, POCDMT000002UK01StructuredBody structuredBody) {
+        List<POCDMT000002UK01Section> sections = getSectionsOfType(structuredBody, ITK_SNOMED, SYSTEM_CODE);
+        for (POCDMT000002UK01Section section : sections) {
+            consent.setSource(new Identifier().setValue(section.getId().getRoot()));
+        }
+    }
+
+    private POCDMT000002UK01StructuredBody getStructuredBody(POCDMT000002UK01ClinicalDocument1 clinicalDocument) {
+        return Optional.ofNullable(clinicalDocument)
+                .map(POCDMT000002UK01ClinicalDocument1::getComponent)
+                .map(POCDMT000002UK01Component2::getStructuredBody)
+                .orElse(null);
+    }
+
+    private List<POCDMT000002UK01Section> getSectionsOfType(POCDMT000002UK01StructuredBody structuredBody,
+                                                            String system, String code) {
+        if (structuredBody == null) return Collections.emptyList();
+
+        return Arrays.stream(structuredBody.getComponentArray())
+                .flatMap(component -> Optional.ofNullable(component.getSection()).stream())
+                .flatMap(section -> Arrays.stream(section.getComponentArray()))
+                .map(POCDMT000002UK01Component5::getSection)
+                .filter(sectionHasCode(system, code))
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    private List<POCDMT000002UK01Entry> getEntriesOfType(POCDMT000002UK01StructuredBody structuredBody,
+                                                         String template) {
+        if (structuredBody == null) return Collections.emptyList();
+
+        return Arrays.stream(structuredBody.getComponentArray())
+                .flatMap(component -> Optional.ofNullable(component.getSection()).stream())
+                .flatMap(section -> Arrays.stream(section.getEntryArray()))
+                .filter(entryHasTemplate(template))
+                .collect(Collectors.toUnmodifiableList());
+    }
+
+    private Predicate<POCDMT000002UK01Section> sectionHasCode(String system, String code) {
+        return section -> Optional.ofNullable(section.getCode())
+                .map(codeElement -> system.equals(codeElement.getCodeSystem())
+                        && code.equals(codeElement.getCode()))
+                .orElse(false);
+    }
+
+    private Predicate<? super POCDMT000002UK01Entry> entryHasTemplate(String template) {
+        return entry -> entry.isSetContentId()
+                && entry.getContentId().getRoot().equals(NPFIT_CDA_CONTENT)
+                && entry.getContentId().getExtension().equals(template);
+    }
+
+    private CodeableConcept getCodingFromCE(CE code) {
+        Coding coding = new Coding(code.getCodeSystem().equalsIgnoreCase(ITK_SNOMED) ?
```

Comment:
code.isSetCodeSystem() -> codeSystem is not mandatory, it could cause NullPointerException here

Reply:
resolved below..